### ISSUE TITLE
Javascript/PHP call?

### ISSUE THREAD

**Issue by jumpjack** [hacf-fr/renault-api#366]
Can anybody provide any hint for porting to javascript or php?
I'd need just the basis, i.e. the payload/header to use for a GET (or POST?) call to get the "full credentials" by inputting the myrenault user/pass couple, and the same for querying the vehicle.
I'm trying to reverse engineering the source code but I have only a very little knowledge of python.

Something that I could test here:
www.curlbuilder.com

or here:
www.reqbin.com/curl

**Comment by oncleben31** [comment]
The starting point of this api in Python was a blog post from @jamesremuscat : https://muscatoxblog.blogspot.com/2019/07/delving-into-renaults-new-api.html?m=1

**Comment by jumpjack** [comment]
That page is really crazy!
And also very confused, and not up to date: we don't need to reverse engineer the app to get the gigya/kamereon credentials, they are contained in public JSON files:

https://renault-wrd-prod-1-euw1-myrapp-one.s3-eu-west-1.amazonaws.com/configuration/android/config_it_IT.json

Part of the response:

```
    "servers": {
        "kamareon": {
            "target": "https://alliance-platform-serviceadapter-staging.apps.prod.eu.kamereon.org",
            "apikey": "Z7P4xyNrTITzNZh52oObSOVfKns9XGLE"
        },
        "wired": {
            "target": "https://api-wired-dev-1-euw1.wrd-aws.com",
            "apikey": "AHdOWFASWEPUVQVlhJWshsios0FqTG2E"
        },
        "wiredValid": {
            "target": "https://api-wired-valid-1-euw1.wrd-aws.com",
            "apikey": "AHdOWFASWEPUVQVlhJWshsios0FqTG2E"
        },
        "wiredProd": {
            "target": "https://api-wired-prod-1-euw1.wrd-aws.com",
            "apikey": "oF09WnKqvBDcrQzcW1rJNpjIuy7KdGaB"
        },
        "gigya": {
            "target": "https://accounts.eu1.gigya.com",
            "apikey": "3_S0OWIrqeJ6mxOkXFT8i3TTDwW1IGKk2rIypZjGXi4Hh8vce2ohuERio1Ka5DBbUr"
        },
        "gigyaValid": {
            "target": "https://accounts.eu1.gigya.com",
            "apikey": "3_7zpQr-4uTyP8xAbB7qK_yTdpNRvNzjw4XKaRnFOkDloiaWf5WkzPxI8ROrchCXvQ"
        },
        "gigyaProd": {
            "target": "https://accounts.eu1.gigya.com",
            "apikey": "3_js8th3jdmCWV86fKR3SXQWvXGKbHoWFv8NAgRbH7FnIBsi_XvCpN_rtLcI07uNuq"
        }
    },
```

Needed data:
Gigya data:
```
        "gigyaProd": {
            "target": "https://accounts.eu1.gigya.com",
            "apikey": "3_js8th3jdmCWV86fKR3SXQWvXGKbHoWFv8NAgRbH7FnIBsi_XvCpN_rtLcI07uNuq"
        }

```
Kamereon data:
```
        "wiredProd": {
            "target": "https://api-wired-prod-1-euw1.wrd-aws.com",
            "apikey": "oF09WnKqvBDcrQzcW1rJNpjIuy7KdGaB"
        },
```

In a  few words:
Gigya apikey:  
`3_js8th3jdmCWV86fKR3SXQWvXGKbHoWFv8NAgRbH7FnIBsi_XvCpN_rtLcI07uNuq`

Kamereon apikey: 
`oF09WnKqvBDcrQzcW1rJNpjIuy7KdGaB`

Using Gigya key we can download "Gigya Session Key", also called in that page "gigya token" or "session token" or "login token" or "oauth token":
https://accounts.eu1.gigya.com/accounts.login?loginID=MAIL&password=PASSWORD&apikey=GIGYA_KEY
(MAIL and PASSWORDS used in myrenault app).

The result is named "cookieValue", which shall be passed to next url:
https://accounts.eu1.gigya.com/accounts.getAccountInfo?apikey=GIGYA_KEY&login_token=COOKIEVALUE

The result is named "PersonId", which we can call "Kamereon Id", in a form like:
`xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx`

**In parallel,** cookieValue is also needed to download "JWT token":
https://accounts.eu1.gigya.com/accounts.getJWT?apikey=GIGYA_KEY&login_token=COOKIEVALUE

The result (we can call it "GIGYA_JWT") is named "Id_token" and is a very long string.

Next step SHOULD be querying Kamereon server using above data, but it is currently not working for me, both using urls or GET/POST queries:

https://api-wired-prod-1-euw1.wrd-aws.com/commerce/v1/persons/PERSONID?apikey=KAMEREON_KEY&x-gigya-id_token=GIGYA_JWT&country=IT

I am using this[ chrome extension](https://chrome.google.com/webstore/detail/restman/ihgpcfpkpmdcghlnaofdmjkoemnlijdi), with no luck, but I am also trying with a javascript/html page, which is not working too because I am using it locally, but I don't want to upload it because it contains my personal data.

-----------

So, in short:

Download api keys **GIGYA_KEY** and **KAMEREON_KEY**:

https://renault-wrd-prod-1-euw1-myrapp-one.s3-eu-west-1.amazonaws.com/configuration/android/config_it_IT.json

Download **Gigya session cookie** (cookieValue):

https://accounts.eu1.gigya.com/accounts.login?loginID=MAIL&password=PASSWORD&apikey=GIGYA_KEY

Download **Kamereon Person Id** (PersonId):

https://accounts.eu1.gigya.com/accounts.getAccountInfo?apikey=GIGYA_KEY&login_token=COOKIEVALUE

Download **Gigya JWT token** (Id_token):

https://accounts.eu1.gigya.com/accounts.getJWT?apikey=GIGYA_KEY&login_token=COOKIEVALUE

Then **query Kamereon server** in some way which I don't understand yet...

To get the account(s) associated to a personId:

https://api-wired-prod-1-euw1.wrd-aws.com/commerce/v1/persons/PERSONID?country=IT

To get vehicles associated to an account:

https://api-wired-prod-1-euw1.wrd-aws.com/commerce/v1/accounts/ACCOUNTID/kamereon/kca/car-adapter/v1/cars/VEHICLE_VIN/location?country=IT

**Comment by epenet** [comment]
The keys are now hard-coded in the renault-api - the ones from AWS are out of date.

**Comment by epenet** [comment]
Direct link: https://github.com/hacf-fr/renault-api/blob/cb92528d32d4d4dbb304ed2d30716159e7642405/src/renault_api/const.py#L21

**Comment by jumpjack** [comment]
Also with new Kamereon Key I keep getting this error:

{"type":"FUNCTIONAL","messages":[{"code":"err.func.wired.unauthorized","message":"The access is unauthorized"}],"errors":[{"errorCode":"err.func.wired.unauthorized","errorMessage":"The access is unauthorized"}],"error_reference":"FUNCTIONAL"}

Used url:

https://api-wired-prod-1-euw1.wrd-aws.com/commerce/v1/persons/MY_PERSON_ID?apikey=Ae9FDWugRxZQAGm3Sxgk7uJn6Q4CGEA2&x-gigya-id_token=MY_JWT_TOKEN

Which is the right url?

**Comment by epenet** [comment]
You are not sharing your code - so it's very hard to help you!
Maybe you are missing some headers.
https://github.com/hacf-fr/renault-api/blob/7c2336c52ed0298a3ed2b8e7dbb77b839676dced/src/renault_api/kamereon/__init__.py#L111-L115

**Comment by jumpjack** [comment]
Because there is no code! I am looking for the raw urls to use. Or, if plain url is not enough for Kamereon part, what to paste into this page:
https://reqbin.com/

The page generates this code upon my current input:

var url = "https://api-wired-prod-1-euw1.wrd-aws.com/commerce/v1/persons/[my person id]?apikey=Ae9FDWugRxZQAGm3Sxgk7uJn6Q4CGEA2&country=IT";

var xhr = new XMLHttpRequest();
xhr.open("GET", url);

xhr.setRequestHeader("x-gigya-id_token", "[my JWT token]");

xhr.onreadystatechange = function () {
   if (xhr.readyState === 4) {
      console.log(xhr.status);
      console.log(xhr.responseText);
   }};

xhr.send();

**Comment by jumpjack** [comment]
Also moving everything into header does not work:

var url = "https://api-wired-prod-1-euw1.wrd-aws.com/commerce/v1/persons/[my person id]?country=IT";

var xhr = new XMLHttpRequest();
xhr.open("GET", url);

xhr.setRequestHeader("x-gigya-id_token", "[my JWT token]");
xhr.setRequestHeader("Content-type", "application/vnd.api+json");
xhr.setRequestHeader("apikey", "Ae9FDWugRxZQAGm3Sxgk7uJn6Q4CGEA2");

xhr.onreadystatechange = function () {
   if (xhr.readyState === 4) {
      console.log(xhr.status);
      console.log(xhr.responseText);
   }};

xhr.send();

**Comment by jumpjack** [comment]
You don't get the point: I am not using that code, it is auto-generated by reqbin site; reqbin site is actually sending my requests using its own internal code based on my input.

**Comment by jumpjack** [comment]
Additional issue:
which is the correct url for an endpoint?

 - https://api-wired-prod-1-euw1.wrd-aws.com/commerce/v1/accounts/MYACCOUNTID/kamereon/kca/car-adapter/v1/cars/MYVIN/actions/hvac-start?country=IT

 - https://api-wired-prod-1-euw1.wrd-aws.com/commerce/v1/accounts/MYACCOUNTID/remote-services/car-adapter/v1/cars/MYVIN/actions/battery-status?country=IT

**Comment by jamesremuscat** [comment]
> That page is really crazy!
> And also very confused, and not up to date

Hi, author of that crazy, confused and not up to date page here (yes, both Gigya and Renault have made changes since I wrote that more than two years ago).

> **In parallel,** cookieValue is also needed to download "JWT token":
> https://accounts.eu1.gigya.com/accounts.getJWT?apikey=GIGYA_KEY&login_token=COOKIEVALUE

This is (part of?) your problem, because this generates a JWT that is useless to Kamereon. You need to specifically request that Gigya includes the fields `data.personId` and `data.gigyaDataCenter` in the JWT it generates. As I wrote 2+ years ago:

> You'll then need to request your Gigya JWT token from accounts.getJWT, again using your Gigya session key as the oauth_token, **and you need to include `data.personId,data.gigyaDataCenter` as the value of `fields`** - Renault's servers need that data to be in the token.

**Comment by jumpjack** [comment]
This is the authentication flow I eventually  reconstructed from a dozen of sources:


> > That page is really crazy!
> > And also very confused, and not up to date
> 
> Hi, author of that crazy, confused and not up to date page here (yes, both Gigya and Renault have made changes since I wrote that more than two years ago).
> 
> > **In parallel,** cookieValue is also needed to download "JWT token":
> > https://accounts.eu1.gigya.com/accounts.getJWT?apikey=GIGYA_KEY&login_token=COOKIEVALUE
> 
> This is (part of?) your problem, because this generates a JWT that is useless to Kamereon. You need to specifically request that Gigya includes the fields `data.personId` and `data.gigyaDataCenter` in the JWT it generates. As I wrote 2+ years ago:
> 
> > You'll then need to request your Gigya JWT token from accounts.getJWT, again using your Gigya session key as the oauth_token, **and you need to include `data.personId,data.gigyaDataCenter` as the value of `fields`** - Renault's servers need that data to be in the token.

This is indeed one of the part not very clear: are they input or output? Now I think they are "inputs requesting an output", i.e. I have to put these inputs to let Gigya understand that I want a specific output in JWT token.
In the meantime, oauth_token changed and became login_token.

And here it is my crazy version of renault authentication method...

![image](https://user-images.githubusercontent.com/1620953/134314518-fd670d62-65aa-4c68-9ba6-2eec50e333ee.png)

I hope I didn't forget anything.

**Comment by jumpjack** [comment]
I eventually succeeded with the final query to the endpoint, but now I have another issue: in Python how do you solve CORS problems? My queries only work if called from an online page, not from my pc, due to "origin: null" header.

**Comment by jumpjack** [comment]
Ok I also solved the CORS issue.
But I noticed that I forgot to mention in the picture the procedure to retrieve the VIN; it is not mandatory but useful, to be able to retrieve it programmatically.

**Comment by epenet** [comment]
Instead of writing all this in the comments, maybe you can provide a documentation PR.
That would be useful in case others need it.

Look at the source on https://github.com/hacf-fr/renault-api/tree/main/docs
Look at the result on https://renault-api.readthedocs.io/en/latest/

**Comment by jumpjack** [comment]
I don't know how to use github.

**Comment by oncleben31** [comment]
But now you have a good reason to learn :-)

To be honest, GitHub is far easier than reverse engineering Renault API.

**Comment by epenet** [comment]
Where is KAMEREON_KEY defined?
Maybe try it like this:
xhr.setRequestHeader("APIkey","Ae9FDWugRxZQAGm3Sxgk7uJn6Q4CGEA2");

**Comment by jumpjack** [comment]
I found that header x-gigya-id_token, and any other header containing "_", is not accepted by my proxy server, for unknown reasons. Any idea?

But it's an "amateur proxy", I copied-pasted it and slightly adapted it for my needs, to solve CORS issues when using my scripts offline.
I added needed headers to this line:
header("Access-Control-Allow-Headers: Content-Type, Access-Control-Allow-Headers, X-Requested-With, X-GIGYA-ID_TOKEN, apikey)
apikey correctly appears in debug output, but X-GIGYA-ID_TOKEN just does not appear in the list of headers received.
Possible alternatives?

**Comment by jumpjack** [comment]
Damn, if Apache is not properly configured, it rejects headers with underscores!

https://stackoverflow.com/questions/17440564/all-caps-http-headers-with-underscores-dropped-in-apache-2-4

**Comment by jumpjack** [comment]
Ok, I eventually solved (or "workarounded") all problems, now I have a working page; it's standalone, you don't need to install anything:

http://win98.altervista.org/myrenault-pub.html

It's still at beta level, so a lot of things on the page are useless for end user.

How to use:
input email and password of MyRenault app/site, and your VIN (which you can find in the app), then click SEND QUERY button; by default the "location" end point is selected, you can paste any of the listed endpoints or test your own.

If you don't trust putting your credentials on an "unknown site", just save the page locally, upload it to your server (together with "myrenault-public.js" file) and use it from there. I didn't yet find a way to get the page working from a local PC.

Currently only reading data is supported, now I have to study the commands...

I'll try uploading the source code to a repo.

**Comment by jumpjack** [comment]
Repository ready:

https://github.com/jumpjack/myRenaultLC

Of course still in very beta stage. Any help is appreciated, Currently I am struggling to retrieve VIN and have password stored by Chrome.
